onnxscript/ir/passes/common/common_subexpression_elimination.py

```diff
@@ -0,0 +1,163 @@
+
+        # 3. Create inputs and initializers in the new graph
+        #    from the old graph.
+        for input in old_graph.inputs:
+            new_input = _copy_value(input)
+            new_graph.inputs.append(new_input)
+            old_value_to_new_value[input] = new_input
+            logger.debug("Adding input %s", new_input.name)
+        for initializer in old_graph.initializers.values():
+            new_initializer = _copy_value(initializer)
+            new_graph.register_initializer(new_initializer)
+            old_value_to_new_value[initializer] = new_initializer
+            logger.debug("Adding initializer %s", new_initializer.name)
+
+        # 4. Create nodes in the new graph from the old graph.
+        for old_node in old_graph:
+            # 4.0. Iterate and update node inputs.
+            old_node_inputs: list[ir.Value] = []
+            for old_input in old_node.inputs:
+                assert old_input is not None
+                if old_input in old_value_to_new_value:
+                    old_node_inputs.append(old_value_to_new_value[old_input])
+                else:
+                    old_node_inputs.append(old_input)
+            # 4.1. Construct the (node, inputs, attributes) hash to
+            #      check if the node is a common subexpression.
+            # Attr is not hashable
+            attributes = {}
+            for k, v in old_node.attributes.items():
+                assert isinstance(v, ir.Attr)
+                attributes[k] = v.value
+            hash_value = hash(
+                (
+                    old_node.op_identifier(),
+                    tuple(old_node_inputs),
+                    tuple(sorted(attributes.items())),
+                )
+            )
+            # TODO(titaiwang): Subgraphs are not supported yet.
+            # TODO(titaiwang): Skip control flow nodes?
+            # 4.2. Check if the node is a common subexpression.
+            if hash_value in old_node_hash_to_new_node:
+                # 4.2.1. If it is, this node is already in the new graph, so
+                #         we don't need to create a new node.
+                modified = True
+                new_node = old_node_hash_to_new_node[hash_value]
+                logger.debug("Reusing node %s", new_node.name)
+            else:
+                # 4.2.2. If it is not, create a new node and add it to the graph.
+                new_node = _copy_node(old_node, old_value_to_new_value)
+                new_graph.append(new_node)
+                old_node_hash_to_new_node[hash_value] = new_node
+            # 4.3 Add the node outputs to the mapping.
+            old_value_to_new_value.update(dict(zip(old_node.outputs, new_node.outputs)))
+        # 5. Create outputs in the new graph from the old graph.
+        for output in old_graph.outputs:
+            new_output = old_value_to_new_value[output]
+            new_graph.outputs.append(new_output)
```

Comment:
We need to ensure that the *names* of outputs (and inputs) of the main model/graph are not changed.

Reply:
This should be taken care of by `ir.convenience.replace_nodes_and_values()`